sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CollationTypeCasts.scala

```diff
@@ -0,0 +1,160 @@
+      if (dt == expr.dataType) expr else Cast(expr, dt)
+    }
+
+  private def castStringType(inType: AbstractDataType, collationId: Int): Option[DataType] = {
+    @Nullable val ret: DataType = inType match {
+      case st: StringType if st.collationId == collationId => st
+      case _: StringType => StringType(collationId)
+      case ArrayType(arrType, nullable) =>
+        castStringType(arrType, collationId).map(ArrayType(_, nullable)).orNull
+      case _ => null
+    }
+    Option(ret)
+  }
+
+  /**
+   * Collates input expressions to a single collation.
+   */
+  def collateToSingleType(exprs: Seq[Expression]): Seq[Expression] = {
+    val collationId = getOutputCollation(exprs)
+
+    exprs.map(e => castStringType(e, collationId).getOrElse(e))
+  }
+
+  /**
+   * Based on the data types of the input expressions this method determines
+   * a collation type which the output will have. This function accepts Seq of
+   * any expressions, but will only be affected by collated StringTypes or
+   * complex DataTypes with collated StringTypes (e.g. ArrayType)
+   */
+  def getOutputCollation(exprs: Seq[Expression]): Int = {
+    val explicitTypes = exprs.filter(hasExplicitCollation)
+      .map(e => extractStringType(e.dataType).collationId).distinct
+
+    explicitTypes.size match {
+      // We have 1 explicit collation
+      case 1 => explicitTypes.head
+      // Multiple explicit collations occurred
+      case size if size > 1 =>
+        throw QueryCompilationErrors
+          .explicitCollationMismatchError(
+            explicitTypes.map(t => StringType(t).typeName)
+          )
+      // Only implicit or default collations present
+      case 0 =>
+        val dataTypes = exprs.filter(e => hasStringType(e.dataType))
+          .map(e => extractStringType(e.dataType))
+
+        if (hasMultipleImplicits(dataTypes)) {
+          throw QueryCompilationErrors.implicitCollationMismatchError()
+        }
+        else {
+          dataTypes.find(dt => !(dt == SQLConf.get.defaultStringType))
+            .getOrElse(SQLConf.get.defaultStringType)
+            .collationId
+        }
+    }
+  }
+
+  /**
+   * This check is always preformed when we have no explicit collation. It returns true
+   * if there are more than one implicit collations. Collations are distinguished by their
+   * collationId.
+   * @param dataTypes
+   * @return
+   */
+  private def hasMultipleImplicits(dataTypes: Seq[StringType]): Boolean =
+    dataTypes.filter(dt => !(dt == SQLConf.get.defaultStringType))
+      .map(_.collationId).distinct.size > 1
+
+  /**
+   * Checks if a given expression has explicitly set collation. For complex DataTypes
+   * we need to check nested children.
+   * @param expression
+   * @return
+   */
+  private def hasExplicitCollation(expression: Expression): Boolean = {
+    expression match {
+      case _: Collate => true
+      case e if e.dataType.isInstanceOf[ArrayType]
+      => expression.children.exists(hasExplicitCollation)
```

Comment:
This looks a bit hacky. What is the exact definition of explicit collation?

Reply:
The exact definition of explicit collation is to if there is a explicit use of COLLATE keyword to set collation of a parameter. In this case, when we have arrays, a query `SELECT array('b' collate unicode_ci) || 'a' collate unicode;` (for us concat_ws) fails in postgreSQL with explicit collation mismatch which requires us to  check nested elements of arrayType for explicitly collated strings.